custom_components/elasticsearch/es_doc_creator.py

```diff
@@ -31,39 +45,64 @@ def __init__(self, hass: HomeAssistant, config: dict) -> None:
     async def async_init(self) -> None:
         """Async initialization."""
 
-        system_info = await self._system_info.async_get_system_info()
         LOGGER.debug("async_init: initializing static doc properties")
+
+        await self._populate_static_doc_properties()
+
+    async def _populate_static_doc_properties(self) -> dict:
         hass_config = self._hass.config
 
-        self._static_doc_properties = {
+        shared_properties = {
             "agent.name": "My Home Assistant",
             "agent.type": "hass",
-            "agent.version": system_info.get("version", "UNKNOWN"),
             "ecs.version": "1.0.0",
             "host.geo.location": {
                 "lat": hass_config.latitude,
                 "lon": hass_config.longitude,
             },
-            "host.architecture": system_info.get("arch", "UNKNOWN"),
-            "host.os.name": system_info.get("os_name", "UNKNOWN"),
-            "host.hostname": system_info.get("hostname", "UNKNOWN"),
-            "tags": self._config.get(CONF_TAGS),
         }
 
-    def state_to_document(self, state: State, time: datetime) -> dict:
-        """Convert entity state to ES document."""
-        try:
-            _state = state_helper.state_as_number(state)
-            if not is_valid_number(_state):
-                _state = state.state
-        except ValueError:
-            _state = state.state
-
-        if time.tzinfo is None:
-            time_tz = time.astimezone(utc)
+        # This is a bit of a hack that allows us to still work if we're passed a mocked config
+        if isinstance(self._config, dict):
+            shared_properties["tags"] = self._config.get(CONF_TAGS, None)
         else:
-            time_tz = time
+            shared_properties["tags"] = None
```

Comment:
This shouldn't be necessary. `self._config` should _always_ be a `dict`. The mock config used in tests is actually a mocked [`ConfigEntry`](https://developers.home-assistant.io/docs/config_entries_index/), which you can think of as a configured instance of the integration. One portion of the `ConfigEntry` is the actual configuration, which is the `dict` being used here.

We have a helper function which extracts the configuration from a `ConfigEntry`, which you can then use to instantiate a `DocumentCreator` class:

https://github.com/legrego/homeassistant-elasticsearch/blob/8e44685708747c69e537339a33634365bc568f87/custom_components/elasticsearch/utils.py#L5-L9

Example usage:
https://github.com/strawgate/homeassistant-elasticsearch/blob/8e44685708747c69e537339a33634365bc568f87/tests/test_es_index_manager.py#L50

Reply:
When I pass a mocked configuration it says that the mocked config object doesn't have a get method.

I'll take a look at using the helper function on documentcreator instantiation!

Reply:
It looks like this works and removes the need for that if statement
```
@pytest.fixture()
async def document_creator(hass: HomeAssistant):
    """Fixture to create a DocumentCreator instance."""

    # Fix the location for the tests
    hass.config.latitude = MOCK_LOCATION_SERVER["lat"]
    hass.config.longitude = MOCK_LOCATION_SERVER["lon"]

    es_url = "http://localhost:9200"
    mock_entry = MockConfigEntry(
        unique_id="test_doc_creator",
        domain=DOMAIN,
        version=3,
        data=build_full_config({"url": es_url, CONF_INDEX_MODE: INDEX_MODE_LEGACY}),
        title="ES Config",
    )

    config = get_merged_config(mock_entry)

    creator = DocumentCreator(hass, config)

    # TODO: Consider initializing the document creator before returning it, requires rewriting tests and initializing the whole integration
    # await creator.async_init()

    yield creator
```

Reply:
> I went back and forth on this so many times haha, I think I don't fully understand how configs work. When I pass a mocked configuration it says that the mocked config object doesn't have a get method.

Yeah, it's confusing. In the beginning, there was only config (`dict`). Then, Home Assistant introduced `ConfigEntry`, which is what enabled integrations to be configured within the UI, as opposed to manually updating `configuration.yaml`.

The config(`dict`) is a portion of the `ConfigEntry`. In this component specifically, it is the combination of the `ConfigEntry`'s `data` and `options` fields, which represent the initial setup config & post-setup customization flows, respectively.